components/webgpu/swapchain.rs

```diff
@@ -343,92 +480,107 @@ impl crate::WGPU {
                     result,
                     global,
                     buffer_id,
-                    buffer_size,
                     wgpu_image_map,
                     context_id,
                     webrender_api,
                     webrender_document,
+                    image_desc,
                 );
             }))
         };
         let map_op = BufferMapOperation {
             host: HostMap::Read,
             callback: Some(callback),
         };
-        let _ = global.buffer_map_async(buffer_id, 0, Some(buffer_size), map_op);
+        global.buffer_map_async(buffer_id, 0, Some(image_desc.buffer_size()), map_op)?;
         self.poller.wake();
-
-        ControlFlow::Continue(())
+        Ok(())
     }
 
-    pub(crate) fn destroy_swapchain(
-        &mut self,
-        context_id: WebGPUContextId,
-        image_key: webrender_api::ImageKey,
-    ) {
-        let present_data = self
-            .wgpu_image_map
+    pub(crate) fn destroy_context(&mut self, context_id: WebGPUContextId) {
+        self.wgpu_image_map
             .lock()
             .unwrap()
             .remove(&context_id)
-            .unwrap();
-        for (buffer_id, buffer_state) in present_data.buffer_ids {
-            match buffer_state {
-                PresentationBufferState::Unassigned => {
-                    /* These buffer were not yet created in wgpu */
-                },
-                _ => {
-                    self.global.buffer_drop(buffer_id);
-                },
-            }
-            if let Err(e) = self.script_sender.send(WebGPUMsg::FreeBuffer(buffer_id)) {
-                warn!("Unable to send FreeBuffer({:?}) ({:?})", buffer_id, e);
-            };
-        }
-        let mut txn = Transaction::new();
-        txn.delete_image(image_key);
-        self.webrender_api
-            .lock()
             .unwrap()
-            .send_transaction(self.webrender_document, txn);
+            .destroy(
+                &self.global,
+                &self.script_sender,
+                &self.webrender_api,
+                self.webrender_document,
+            );
     }
 }
 
 fn update_wr_image(
     result: Result<(), BufferAccessError>,
     global: Arc<Global>,
     buffer_id: id::BufferId,
-    buffer_size: u64,
     wgpu_image_map: WGPUImageMap,
     context_id: WebGPUContextId,
     webrender_api: Arc<Mutex<RenderApi>>,
     webrender_document: webrender_api::DocumentId,
+    image_desc: WebGPUImageDescriptor,
 ) {
     match result {
         Ok(()) => {
-            if let Some(present_data) = wgpu_image_map.lock().unwrap().get_mut(&context_id) {
-                let buffer_state = present_data.get_buffer_state(buffer_id);
-                assert_eq!(*buffer_state, PresentationBufferState::Mapping);
+            if let Some(context_data) = wgpu_image_map.lock().unwrap().get_mut(&context_id) {
+                let config_changed = image_desc != context_data.image_desc;
```

Comment:
This checks for any change in image_desc (in practice only format, flags and size changes with update context for now) and as noted below this is not reliable (we can be reconfigured to same config) in any case this mapasync call is outdated.

> As I understand it, the problems comes from replace_drawing_buffer doing send_swap_chain_present followed by update_context, and the fact that update_wr_image could be called into either before or after update_context. Sounds right?

Yes

> I am not sure what you mean, but I think using a state machine for the processing steps of the descriptor(s)(?) would allow you to cover all cases without having to block.

I do not think we can actually model this with state machines, except with something like submission indexes:

this would be something like it https://docs.rs/wgpu/latest/wgpu/struct.SubmissionIndex.html, basically we number each mapasync call with number (inc for each call), then when doing map async callback we make sure to only use newer data. We could also note last submission index for current config, so all mapasync from older configs get thrown away correctly.

Reply:
Ok sounds good, in this PR or a follow-up?

Reply:
let's do in follow up.